bolt-modules/boltlib/lib/puppet/functions/wait.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+require 'bolt/logger'
+require 'bolt/target'
+
+# Wait for a Future or array of Futures to finish and return results.
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:wait, Puppet::Functions::InternalFunction) do
+  # Wait for a Future or array of Futures to finish and return results.
+  # @param futures A Bolt Future object or array of Bolt Futures to wait on.
+  # @option options [Boolean] _catch_errors Whether to catch raised errors.
+  # @return A Result or Results from the Futures
+  # @example Upload a large file in the background, then wait until it's loaded
+  #   $futures = background() || {
+  #     upload_file("./very_large_file", "/opt/jfrog/artifactory/var/etc/artifactory", $targets)
+  #   }
+  #   # Run an unrelated task
+  #   run_task("deploy", $targets)
+  #   # Wait for the file upload to finish
+  #   $results = wait($futures)
+  dispatch :wait do
+    param 'Variant[Future, Array[Future]]', :futures
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'Array[Boltlib::PlanResult]'
+  end
+
+  # Wait for a Future or array of Futures to finish and return results.
+  # @param futures A Bolt Future object or array of Bolt Futures to wait on.
+  # @param timeout How long to wait for Futures to finish before raising a Timeout error.
+  # @option options [Boolean] _catch_errors Whether to catch raised errors.
+  # @return A Result or Results from the Futures
+  # @example Upload a large file in the background, then wait until it's loaded
+  #   $futures = background() || {
+  #     upload_file("./very_large_file", "/opt/jfrog/artifactory/var/etc/artifactory", $targets)
+  #   }
+  #   # Run an unrelated task
+  #   run_task("deploy", $targets)
+  #   # Wait for the file upload to finish
+  #   $results = wait($futures)
+  dispatch :wait_with_timeout do
+    param 'Variant[Future, Array[Future]]', :futures
+    param 'Variant[Integer[0], Float[0.0]]', :timeout
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'Array[Boltlib::PlanResult]'
+  end
+
+  def wait(futures, options = {})
+    wait_with_timeout(futures, nil, options)
+  end
+
+  def wait_with_timeout(futures, timeout, options = {})
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue
+        .from_issue_and_stack(Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, action: 'wait')
+    end
+
+    valid, invalid = options.partition { |k, _v| %w[_catch_errors].include?(k) }.map(&:to_h)
+    if invalid.any?
+      file, line = Puppet::Pops::PuppetStack.top_of_stack
+      msg = "The wait() function call in #{file}#L#{line} received invalid options "\
+        "#{invalid.keys}. Removing invalid options and continuing..."
+      Bolt::Logger.warn("wait_function_options", msg)
+    end
```

Comment:
I don't think we have this pattern anywhere else, at least in plan functions. I think we should either just silently ignore unknown options or use a more general warning ID like `plan_function_options`.

Reply:
Yeah, we don't, but it seemed like it could be useful? I can remove it if we want to add this to more functions as it's own ticket (this PR is certainly big enough :P)

Reply:
Ya, I don't mind having it at all (as long as the ID is more generic). Your call if it goes in this ticket or another one!

Reply:
Definitely - updated the ID, but leaving this here.